README.md
shadowsocks-juice-mu
===========



shadowsocks
===========

[![PyPI version]][PyPI]
[![Build Status]][Travis CI]

A fast tunnel proxy that helps you bypass firewalls.

Features:
- TCP & UDP support
- User management API
- TCP Fast Open
- Workers and graceful restart
- Destination IP blacklist

Server
------

### Install

Debian / Ubuntu:

    apt-get install python-pip
    pip install git+https://github.com/shadowsocks/shadowsocks.git@master

CentOS:

    yum install python-setuptools && easy_install pip
    pip install git+https://github.com/shadowsocks/shadowsocks.git@master

Windows:

See [Install Shadowsocks Server on Windows](https://github.com/shadowsocks/shadowsocks/wiki/Install-Shadowsocks-Server-on-Windows).

### Usage

    ssserver -p 443 -k password -m aes-256-cfb

To run in the background:

    sudo ssserver -p 443 -k password -m aes-256-cfb --user nobody -d start

To stop:

    sudo ssserver -d stop

To check the log:

    sudo less /var/log/shadowsocks.log

Check all the options via `-h`. You can also use a [Configuration] file
instead.

### Usage with Config File

[Create configeration file and run](https://github.com/shadowsocks/shadowsocks/wiki/Configuration-via-Config-File)

To start:

    ssserver -c /etc/shadowsocks.json


Documentation
-------------

You can find all the documentation in the [Wiki](https://github.com/shadowsocks/shadowsocks/wiki).

License
-------

Apache License







[Build Status]:      https://img.shields.io/travis/shadowsocks/shadowsocks/master.svg?style=flat
[PyPI]:              https://pypi.python.org/pypi/shadowsocks
[PyPI version]:      https://img.shields.io/pypi/v/shadowsocks.svg?style=flat
[Travis CI]:         https://travis-ci.org/shadowsocks/shadowsocks



CHANGES
2.8.2 2015-08-10
- Fix a encoding problem in manager

2.8.1 2015-08-06
- Respond ok to add and remove commands

2.8 2015-08-06
- Add Shadowsocks manager

2.7 2015-08-02
- Optimize speed for multiple ports

2.6.11 2015-07-10
- Fix a compatibility issue in UDP Relay

2.6.10 2015-06-08
- Optimize LRU cache
- Refine logging

2.6.9 2015-05-19
- Fix a stability issue on Windows

2.6.8 2015-02-10
- Support multiple server ip on client side
- Support --version
- Minor fixes

2.6.7 2015-02-02
- Support --user
- Support CIDR format in --forbidden-ip
- Minor fixes

2.6.6 2015-01-23
- Fix a crash in forbidden list

2.6.5 2015-01-18
- Try both 32 bit and 64 bit dll on Windows

2.6.4 2015-01-14
- Also search lib* when searching libraries

2.6.3 2015-01-12
- Support --forbidden-ip to ban some IP, i.e. localhost
- Search OpenSSL and libsodium harder
- Now works on OpenWRT

2.6.2 2015-01-03
- Log client IP

2.6.1 2014-12-26
- Fix a problem with TCP Fast Open on local side
- Fix sometimes daemon_start returns wrong exit status

2.6 2014-12-21
- Add daemon support

2.5 2014-12-11
- Add salsa20 and chacha20

2.4.3 2014-11-10
- Fix an issue on Python 3
- Fix an issue with IPv6

2.4.2 2014-11-06
- Fix command line arguments on Python 3
- Support table on Python 3
- Fix TCP Fast Open on Python 3

2.4.1 2014-11-01
- Fix setup.py for non-utf8 locales on Python 3

2.4 2014-11-01
- Python 3 support
- Performance improvement
- Fix LRU cache behavior

2.3.2 2014-10-11
- Fix OpenSSL on Windows

2.3.1 2014-10-09
- Does not require M2Crypto any more

2.3 2014-09-23
- Support CFB1, CFB8 and CTR mode of AES
- Do not require password config when using port_password
- Use SIGTERM instead of SIGQUIT on Windows

2.2.2 2014-09-14
- Fix when multiple DNS set, IPv6 only sites are broken

2.2.1 2014-09-10
- Support graceful shutdown
- Fix some bugs

2.2.0 2014-09-09
- Add RC4-MD5 encryption

2.1.0 2014-08-10
- Use only IPv4 DNS server
- Does not ship config.json
- Better error message

2.0.12 2014-07-26
- Support -q quiet mode
- Exit 0 when showing help with -h

2.0.11 2014-07-12
- Prefers IP addresses over hostnames, more friendly with socksify and openvpn

2.0.10 2014-07-11
- Fix UDP on local

2.0.9 2014-07-06
- Fix EWOULDBLOCK on Windows
- Fix Unicode config problem on some platforms

2.0.8 2014-06-23
- Use multiple DNS to query hostnames

2.0.7 2014-06-21
- Fix fastopen on local
- Fallback when fastopen is not available
- Add verbose logging mode -vv
- Verify if hostname is valid

2.0.6 2014-06-19
- Fix CPU 100% on POLL_HUP
- More friendly logging

2.0.5 2014-06-18
- Support a simple config format for multiple ports

2.0.4 2014-06-12
- Fix worker master

2.0.3 2014-06-11
- Fix table encryption with UDP

2.0.2 2014-06-11
- Add asynchronous DNS in TCP relay

2.0.1 2014-06-05
- Better logging
- Maybe fix bad file descriptor

2.0 2014-06-05
- Use a new event model
- Remove gevent
- Refuse to use default password
- Fix a problem when using multiple passwords with table encryption

1.4.5 2014-05-24
- Add timeout in TCP server
- Close sockets in master process

1.4.4 2014-05-17
- Support multiple workers

1.4.3 2014-05-13
- Fix Windows

1.4.2 2014-05-10
- Add salsa20-ctr cipher

1.4.1 2014-05-03
- Fix error log
- Fix EINPROGESS with some version of gevent

1.4.0 2014-05-02
- Adds UDP relay
- TCP fast open support on Linux 3.7+

1.3.7 2014-04-10
- Fix a typo in help

1.3.6 2014-04-10
- Fix a typo in help

1.3.5 2014-04-07
- Add help
- Change default local binding address into 127.0.0.1

1.3.4 2014-02-17
- Fix a bug when no config file exists
- Client now support multiple server ports and multiple server/port pairs
- Better error message with bad config.json format and wrong password

1.3.3 2013-07-09
- Fix default key length of rc2

1.3.2 2013-07-04
- Server will listen at server IP specified in config
- Check config file and show some warning messages

1.3.1 2013-06-29
- Fix -c arg

1.3.0 2013-06-22
- Move to pypi

1.2.3 2013-06-14
- add bind address

1.2.2 2013-05-31
- local can listen at ::0 with -6 arg; bump 1.2.2

1.2.1 2013-05-23
- Fix an OpenSSL crash

1.2 2013-05-22
- Use random iv, we finally have strong encryption

1.1.1 2013-05-21
- Add encryption, AES, blowfish, etc.

1.1 2013-05-16
- Support IPv6 addresses (type 4)
- Drop Python 2.5 support

1.0 2013-04-03
- Fix -6 IPv6

0.9.4 2013-03-04
- Support Python 2.5

0.9.3 2013-01-14
- Fix conn termination null data

0.9.2 2013-01-05
- Change default timeout

0.9.1 2013-01-05
- Add Travis-CI test

0.9 2012-12-30
- Replace send with sendall, fix FreeBSD

0.6 2012-12-06
- Support args

0.5 2012-11-08
- Fix encryption with negative md5sum

0.4 2012-11-02
- Move config into a JSON file
- Auto-detect config path

0.3 2012-06-06
- Move socks5 negotiation to local

0.2 2012-05-11
- Add -6 arg for IPv6
- Fix socket.error

0.1 2012-04-20
- Initial version


shadowsocks/crypto/sodium.py
#!/usr/bin/env python
#
# Copyright 2015 clowwindy
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from __future__ import absolute_import, division, print_function, \
    with_statement

from ctypes import c_char_p, c_int, c_uint, c_ulonglong, byref, \
    create_string_buffer, c_void_p

from shadowsocks.crypto import util
from shadowsocks.crypto import aead
from shadowsocks.crypto.aead import AeadCryptoBase

__all__ = ['ciphers']

libsodium = None
loaded = False

buf = None
buf_size = 2048

# for salsa20 and chacha20 and chacha20-ietf
BLOCK_SIZE = 64


def load_libsodium(crypto_path=None):
    global loaded, libsodium, buf

    crypto_path = dict(crypto_path) if crypto_path else dict()
    path = crypto_path.get('sodium', None)

    if not aead.sodium_loaded:
        aead.load_sodium(path)

    if aead.sodium_loaded:
        libsodium = aead.libsodium
    else:
        print('load libsodium again with path %s' % path)
        libsodium = util.find_library('sodium', 'crypto_stream_salsa20_xor_ic',
                                      'libsodium', path)
        if libsodium is None:
            raise Exception('libsodium not found')

        if libsodium.sodium_init() < 0:
            raise Exception('libsodium init failed')

    libsodium.crypto_stream_salsa20_xor_ic.restype = c_int
    libsodium.crypto_stream_salsa20_xor_ic.argtypes = (
        c_void_p, c_char_p,  # cipher output, msg
        c_ulonglong,  # msg len
        c_char_p, c_ulonglong,  # nonce, uint64_t initial block counter
        c_char_p  # key
    )
    libsodium.crypto_stream_chacha20_xor_ic.restype = c_int
    libsodium.crypto_stream_chacha20_xor_ic.argtypes = (
        c_void_p, c_char_p,
        c_ulonglong,
        c_char_p, c_ulonglong,
        c_char_p
    )
    if hasattr(libsodium, 'crypto_stream_xchacha20_xor_ic'):
        libsodium.crypto_stream_xchacha20_xor_ic.restype = c_int
        libsodium.crypto_stream_xchacha20_xor_ic.argtypes = (
            c_void_p, c_char_p,
            c_ulonglong,
            c_char_p, c_ulonglong,
            c_char_p
        )
    libsodium.crypto_stream_chacha20_ietf_xor_ic.restype = c_int
    libsodium.crypto_stream_chacha20_ietf_xor_ic.argtypes = (
        c_void_p, c_char_p,
        c_ulonglong,
        c_char_p,
        c_uint,  # uint32_t initial counter
        c_char_p
    )

    # chacha20-poly1305
    libsodium.crypto_aead_chacha20poly1305_encrypt.restype = c_int
    libsodium.crypto_aead_chacha20poly1305_encrypt.argtypes = (
        c_void_p, c_void_p,  # c, clen
        c_char_p, c_ulonglong,  # m, mlen
        c_char_p, c_ulonglong,  # ad, adlen
        c_char_p,  # nsec, not used
        c_char_p, c_char_p  # npub, k
    )
    libsodium.crypto_aead_chacha20poly1305_decrypt.restype = c_int
    libsodium.crypto_aead_chacha20poly1305_decrypt.argtypes = (
        c_void_p, c_void_p,  # m, mlen
        c_char_p,  # nsec, not used
        c_char_p, c_ulonglong,  # c, clen
        c_char_p, c_ulonglong,  # ad, adlen
        c_char_p, c_char_p  # npub, k
    )

    # chacha20-ietf-poly1305, same api structure as above
    libsodium.crypto_aead_chacha20poly1305_ietf_encrypt.restype = c_int
    libsodium.crypto_aead_chacha20poly1305_ietf_encrypt.argtypes = (
        c_void_p, c_void_p,
        c_char_p, c_ulonglong,
        c_char_p, c_ulonglong,
        c_char_p,
        c_char_p, c_char_p
    )
    libsodium.crypto_aead_chacha20poly1305_ietf_decrypt.restype = c_int
    libsodium.crypto_aead_chacha20poly1305_ietf_decrypt.argtypes = (
        c_void_p, c_void_p,
        c_char_p,
        c_char_p, c_ulonglong,
        c_char_p, c_ulonglong,
        c_char_p, c_char_p
    )

    # xchacha20-ietf-poly1305, same api structure as above
    if hasattr(libsodium, 'crypto_aead_xchacha20poly1305_ietf_encrypt'):
        libsodium.crypto_aead_xchacha20poly1305_ietf_encrypt.restype = c_int
        libsodium.crypto_aead_xchacha20poly1305_ietf_encrypt.argtypes = (
            c_void_p, c_void_p,
            c_char_p, c_ulonglong,
            c_char_p, c_ulonglong,
            c_char_p,
            c_char_p, c_char_p
        )

        libsodium.crypto_aead_xchacha20poly1305_ietf_decrypt.restype = c_int
        libsodium.crypto_aead_xchacha20poly1305_ietf_decrypt.argtypes = (
            c_void_p, c_void_p,
            c_char_p,
            c_char_p, c_ulonglong,
            c_char_p, c_ulonglong,
            c_char_p, c_char_p
        )

    # aes-256-gcm, same api structure as above
    libsodium.crypto_aead_aes256gcm_is_available.restype = c_int

    if libsodium.crypto_aead_aes256gcm_is_available():
        libsodium.crypto_aead_aes256gcm_encrypt.restype = c_int
        libsodium.crypto_aead_aes256gcm_encrypt.argtypes = (
            c_void_p, c_void_p,
            c_char_p, c_ulonglong,
            c_char_p, c_ulonglong,
            c_char_p,
            c_char_p, c_char_p
        )
        libsodium.crypto_aead_aes256gcm_decrypt.restype = c_int
        libsodium.crypto_aead_aes256gcm_decrypt.argtypes = (
            c_void_p, c_void_p,
            c_char_p,
            c_char_p, c_ulonglong,
            c_char_p, c_ulonglong,
            c_char_p, c_char_p
        )

    buf = create_string_buffer(buf_size)
    loaded = True


class SodiumCrypto(object):
    def __init__(self, cipher_name, key, iv, op, crypto_path=None):
        if not loaded:
            load_libsodium(crypto_path)
        self.key = key
        self.iv = iv
        self.key_ptr = c_char_p(key)
        self.iv_ptr = c_char_p(iv)
        if cipher_name == 'salsa20':
            self.cipher = libsodium.crypto_stream_salsa20_xor_ic
        elif cipher_name == 'chacha20':
            self.cipher = libsodium.crypto_stream_chacha20_xor_ic
        elif cipher_name == 'xchacha20':
            if hasattr(libsodium, 'crypto_stream_xchacha20_xor_ic'):
                self.cipher = libsodium.crypto_stream_xchacha20_xor_ic
            else:
                raise Exception('Unsupported cipher')
        elif cipher_name == 'chacha20-ietf':
            self.cipher = libsodium.crypto_stream_chacha20_ietf_xor_ic
        else:
            raise Exception('Unknown cipher')
        # byte counter, not block counter
        self.counter = 0
        self.encrypt = self.update
        self.decrypt = self.update
        self.encrypt_once = self.update
        self.decrypt_once = self.update

    def update(self, data):
        global buf_size, buf
        l = len(data)

        # we can only prepend some padding to make the encryption align to
        # blocks
        padding = self.counter % BLOCK_SIZE
        if buf_size < padding + l:
            buf_size = (padding + l) * 2
            buf = create_string_buffer(buf_size)

        if padding:
            data = (b'\0' * padding) + data
        self.cipher(byref(buf), c_char_p(data), padding + l,
                    self.iv_ptr, int(self.counter / BLOCK_SIZE), self.key_ptr)
        self.counter += l
        # buf is copied to a str object when we access buf.raw
        # strip off the padding
        return buf.raw[padding:padding + l]

    def clean(self):
        pass


class SodiumAeadCrypto(AeadCryptoBase):
    def __init__(self, cipher_name, key, iv, op, crypto_path=None):
        if not loaded:
            load_libsodium(crypto_path)
        AeadCryptoBase.__init__(self, cipher_name, key, iv, op, crypto_path)

        if cipher_name == 'chacha20-poly1305':
            self.encryptor = libsodium.crypto_aead_chacha20poly1305_encrypt
            self.decryptor = libsodium.crypto_aead_chacha20poly1305_decrypt
        elif cipher_name == 'chacha20-ietf-poly1305':
            self.encryptor = libsodium. \
                crypto_aead_chacha20poly1305_ietf_encrypt
            self.decryptor = libsodium. \
                crypto_aead_chacha20poly1305_ietf_decrypt
        elif cipher_name == 'xchacha20-ietf-poly1305':
            if hasattr(libsodium,
                       'crypto_aead_xchacha20poly1305_ietf_encrypt'):
                self.encryptor = libsodium. \
                    crypto_aead_xchacha20poly1305_ietf_encrypt
                self.decryptor = libsodium. \
                    crypto_aead_xchacha20poly1305_ietf_decrypt
            else:
                raise Exception('Unsupported cipher')
        elif cipher_name == 'sodium:aes-256-gcm':
            if hasattr(libsodium, 'crypto_aead_aes256gcm_encrypt'):
                self.encryptor = libsodium.crypto_aead_aes256gcm_encrypt
                self.decryptor = libsodium.crypto_aead_aes256gcm_decrypt
            else:
                raise Exception('Unsupported cipher')
        else:
            raise Exception('Unknown cipher')

    def cipher_ctx_init(self):
        global libsodium
        libsodium.sodium_increment(byref(self._nonce), c_int(self._nlen))
        # print("".join("%02x" % ord(b) for b in self._nonce))

    def aead_encrypt(self, data):
        global buf, buf_size
        plen = len(data)
        if buf_size < plen + self._tlen:
            buf_size = (plen + self._tlen) * 2
            buf = create_string_buffer(buf_size)
        cipher_out_len = c_ulonglong(0)
        self.encryptor(
            byref(buf), byref(cipher_out_len),
            c_char_p(data), c_ulonglong(plen),
            None, c_ulonglong(0), None,
            c_char_p(self._nonce.raw), c_char_p(self._skey)
        )
        if cipher_out_len.value != plen + self._tlen:
            raise Exception("Encrypt failed")

        self.cipher_ctx_init()
        return buf.raw[:cipher_out_len.value]

    def aead_decrypt(self, data):
        global buf, buf_size
        clen = len(data)
        if buf_size < clen:
            buf_size = clen * 2
            buf = create_string_buffer(buf_size)
        cipher_out_len = c_ulonglong(0)
        r = self.decryptor(
            byref(buf), byref(cipher_out_len),
            None,
            c_char_p(data), c_ulonglong(clen),
            None, c_ulonglong(0),
            c_char_p(self._nonce.raw), c_char_p(self._skey)
        )
        if r != 0:
            raise Exception("Decrypt failed")

        if cipher_out_len.value != clen - self._tlen:
            raise Exception("Decrypt failed")

        self.cipher_ctx_init()
        return buf.raw[:cipher_out_len.value]


ciphers = {
    'salsa20': (32, 8, SodiumCrypto),
    'chacha20': (32, 8, SodiumCrypto),
    'xchacha20': (32, 24, SodiumCrypto),
    'chacha20-ietf': (32, 12, SodiumCrypto),
    # AEAD: iv_len = salt_len = key_len
    'chacha20-poly1305': (32, 32, SodiumAeadCrypto),
    'chacha20-ietf-poly1305': (32, 32, SodiumAeadCrypto),
    'xchacha20-ietf-poly1305': (32, 32, SodiumAeadCrypto),
    'sodium:aes-256-gcm': (32, 32, SodiumAeadCrypto),
}


def test_chacha20():
    print("Test chacha20")
    cipher = SodiumCrypto('chacha20', b'k' * 32, b'i' * 16, 1)
    decipher = SodiumCrypto('chacha20', b'k' * 32, b'i' * 16, 0)

    util.run_cipher(cipher, decipher)


def test_xchacha20():
    print("Test xchacha20")
    cipher = SodiumCrypto('xchacha20', b'k' * 32, b'i' * 24, 1)
    decipher = SodiumCrypto('xchacha20', b'k' * 32, b'i' * 24, 0)

    util.run_cipher(cipher, decipher)


def test_salsa20():
    print("Test salsa20")
    cipher = SodiumCrypto('salsa20', b'k' * 32, b'i' * 16, 1)
    decipher = SodiumCrypto('salsa20', b'k' * 32, b'i' * 16, 0)

    util.run_cipher(cipher, decipher)


def test_chacha20_ietf():
    print("Test chacha20-ietf")
    cipher = SodiumCrypto('chacha20-ietf', b'k' * 32, b'i' * 16, 1)
    decipher = SodiumCrypto('chacha20-ietf', b'k' * 32, b'i' * 16, 0)

    util.run_cipher(cipher, decipher)


def test_chacha20_poly1305():
    print("Test chacha20-poly1305 [payload][tag]")
    cipher = SodiumAeadCrypto('chacha20-poly1305',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('chacha20-poly1305',
                                b'k' * 32, b'i' * 32, 0)

    util.run_cipher(cipher, decipher)


def test_chacha20_poly1305_chunk():
    print("Test chacha20-poly1305 chunk [size][tag][payload][tag]")
    cipher = SodiumAeadCrypto('chacha20-poly1305',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('chacha20-poly1305',
                                b'k' * 32, b'i' * 32, 0)

    cipher.encrypt_once = cipher.encrypt
    decipher.decrypt_once = decipher.decrypt

    util.run_cipher(cipher, decipher)


def test_chacha20_ietf_poly1305():
    print("Test chacha20-ietf-poly1305 [payload][tag]")
    cipher = SodiumAeadCrypto('chacha20-ietf-poly1305',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('chacha20-ietf-poly1305',
                                b'k' * 32, b'i' * 32, 0)

    util.run_cipher(cipher, decipher)


def test_chacha20_ietf_poly1305_chunk():
    print("Test chacha20-ietf-poly1305 chunk [size][tag][payload][tag]")
    cipher = SodiumAeadCrypto('chacha20-ietf-poly1305',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('chacha20-ietf-poly1305',
                                b'k' * 32, b'i' * 32, 0)

    cipher.encrypt_once = cipher.encrypt
    decipher.decrypt_once = decipher.decrypt

    util.run_cipher(cipher, decipher)


def test_aes_256_gcm():
    print("Test sodium:aes-256-gcm [payload][tag]")
    cipher = SodiumAeadCrypto('sodium:aes-256-gcm',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('sodium:aes-256-gcm',
                                b'k' * 32, b'i' * 32, 0)

    util.run_cipher(cipher, decipher)


def test_aes_256_gcm_chunk():
    print("Test sodium:aes-256-gcm chunk [size][tag][payload][tag]")
    cipher = SodiumAeadCrypto('sodium:aes-256-gcm',
                              b'k' * 32, b'i' * 32, 1)
    decipher = SodiumAeadCrypto('sodium:aes-256-gcm',
                                b'k' * 32, b'i' * 32, 0)

    cipher.encrypt_once = cipher.encrypt
    decipher.decrypt_once = decipher.decrypt

    util.run_cipher(cipher, decipher)


if __name__ == '__main__':
    test_chacha20()
    test_xchacha20()
    test_salsa20()
    test_chacha20_ietf()
    test_chacha20_poly1305()
    test_chacha20_poly1305_chunk()
    test_chacha20_ietf_poly1305()
    test_chacha20_ietf_poly1305_chunk()
    test_aes_256_gcm()
    test_aes_256_gcm_chunk()
